### DISCUSSION TITLE
About the problem of software restart of STM32duino

### DISCUSSION THREAD

**Discussion by JingtaiChen** [stm32duino/Arduino_Core_STM32#1909] [Q&A] [answered]
About the problem of software restart of STM32duino
I have two questions as follows:
1. When the stm32 accidentally crashes due to the single use of stm32, the stm32 will only be stuck there, while the 328p will automatically restart.
2. When I artificially want to restart stm32, generally speaking, I will write the following program on arduino. With this program, 328p will restart, while stm32 will be stuck there and will not restart.
Looking forward to your reply, I wish all the best!
`#include <Arduino.h>
void(* resetFunc) (void) = 0;//定义重启变量

void setup()
{
Serial.begin(115200);
Serial.println("启动成功");
delay(1000);
}

void loop()
{
Serial.println("重启");
delay(1000);
resetFunc();//重启
}`

**Comment by fpistm** [comment]
There is no issue. STM32 Simply fallback to an error handler. And often it ends with a `while(1)`.
Try 
`NVIC_SystemReset();`

Most of those error handler are weak and can be redefined.

**Comment by JingtaiChen** [comment]
Thank you for your reply, your method is very effective! So if I want the program to be able to restart itself when it crashes, I need to write my own watchdog program?

**Reply by fpistm** [reply]
Yes. You can redefine them as they are defined weak.

**Reply by JingtaiChen** [reply]
I wonder how to do that?

**Reply by fpistm** [reply]
And please avoid to open issue for question. Use the forum or GitHub  discussion. But you will get better audience on the forum.

**Reply by JingtaiChen** [reply]
ok
thank you for your reply

**Reply by fpistm** [reply, marked as answer]
Or use this library provided within the core:
https://github.com/stm32duino/Arduino_Core_STM32/tree/main/libraries/IWatchdog
